### PR TITLE
Support Iceberg table format

### PR DESCRIPTION
**Pull request by andrei-ionescu**
### What is the context for this pull request?

 - **Tracking Issue**: #306
 - **Proposal**: #318
 - **Dependencies**: #355 
 - Fixes: #306
 - Fixes: #318
 - Closes: #321 
 - Closes: #320 

### What changes were proposed in this pull request?

This PR adds support for Iceberg.

The following changes are in this PR and each of them are separate commits:

- Add Iceberg source.
- Add support for incremental refresh. This is based on #301 PR from @sezruby.
- Add integration test

### Does this PR introduce _any_ user-facing change?

No. The main changes to user-facing APIs are in the #321 PR. Detailed information can be found in the #318 proposal.

### How was this patch tested?

1. Integration test added for the new functionality
2. Locally & Databricks Runtime tests

- Local build

```
sbt publishLocal
```

- Run Spark shell with Hyperspace and Iceberg libraries loaded

```
$ spark-shell \
--driver-memory 4g \
--packages "com.microsoft.hyperspace:hyperspace-core_2.11:0.4.0-SNAPSHOT,org.apache.iceberg:iceberg-spark-runtime:0.10.0" \
--driver-java-options "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5006 -XX:+UseG1GC -Dlog4j.debug=true"
```

- Paste the following code

```
import org.apache.spark.sql._
import com.microsoft.hyperspace._
import com.microsoft.hyperspace.index._
import scala.collection.JavaConverters._
import org.apache.iceberg.PartitionSpec
import org.apache.iceberg.TableProperties
import org.apache.iceberg.spark._
import org.apache.iceberg.hadoop._

val hs = new Hyperspace(spark)

// create Iceberg table
val props = Map(TableProperties.WRITE_NEW_DATA_LOCATION -> "table3").asJava
val sourceDf = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
val schema = SparkSchemaUtil.convert(sourceDf.schema)
val part = PartitionSpec.builderFor(schema).build()
val icebergTable = new HadoopTables().create(schema, part, props, "table3")
sourceDf.write.mode("overwrite").format("iceberg").save("./table3")

// read created table
val iceDf = spark.read.format("iceberg").load("./table3")

// create indexes
hs.createIndex(iceDf, IndexConfig("index_ice0", indexedColumns = Seq("id"), includedColumns = Seq("name")))
hs.createIndex(iceDf, IndexConfig("index_ice1", indexedColumns = Seq("name")))

// verify plans
val query = iceDf.filter(iceDf("id") === 1).select("name")
hs.explain(query, verbose = true)
```

### CODE REVIEW

**Comment by andrei-ionescu** [comment]
@imback82 I closed the #321, #320 PRs which were obsolete after your refactoring work on #355 - thanks for it. Please review this PR that adds support for Iceberg table format.

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:218]: Could you run `scalafmt` in this file?

**Review by imback82 — COMMENTED** [review]
Few minor/nit comments, but generally looking good to me.

I don't know every detail of Iceberg, but since the changes are self-contained, I think this is good to go. @sezruby could you also take a look? Thanks.

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:218]: and `IcebergIntegrationTest.scala` as well.

**Review by sezruby — COMMENTED** [review]
Generally looks good to me! Could you also consider adding `HybridScanForIcebergTest` like `HybridScanForDeltaLakeTest`?

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (andrei-ionescu) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:277]: Renamed to `path`.

**Comment by andrei-ionescu** [comment]
@sezruby The test for hybrid scan is in the `IcebergIntegrationTest` lines [299-338](https://github.com/microsoft/hyperspace/pull/358/files#diff-ce1f32f296e1683385beb0fe1954b154710c0ba0120f028167afbe5953347dd3R299-R338).

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:195]: nit: can this be just `private lazy val icebergTable: Table` so that we don't have to do `val t = getIcebergTable` on the caller side?

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:237]: nit: we don't specify types (unless required) for local variables.
```suggestion
    var scan = table.newScan.caseSensitive(caseSensitive).project(lazySchema)
```

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergFileBasedSource.scala
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergFileBasedSource.scala:68]: let's update delta

**Review by imback82 — COMMENTED** [review]
LGTM (pending minor/nit comments + @sezruby's comments), thanks @andrei-ionescu!

**Comment by sezruby** [comment]
> @sezruby The test for hybrid scan is in the IcebergIntegrationTest lines 299-338.

Yes, but we need to verify the exact plan transformation and the result of query.

**Comment by andrei-ionescu** [comment]
@sezruby I did take the Delta Lake hybrid test and modified it for Iceberg. That is what is in the integration test from line 299 to 338. Is there other hybrid test for Delta Lake that I missed?

**Comment by sezruby** [comment]
@andrei-ionescu There's another test set:
https://github.com/microsoft/hyperspace/blob/master/src/test/scala/com/microsoft/hyperspace/index/HybridScanForDeltaLakeTest.scala 
based on https://github.com/microsoft/hyperspace/blob/master/src/test/scala/com/microsoft/hyperspace/index/HybridScanSuite.scala

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (andrei-ionescu) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:195]: There are some Iceberg options that may change from other writers and in this case it is desired to get the latest table instance.

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:158]: not used?

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:195]: ah ok, make sense. Shall we rename this to `loadIcebergTable` to be explicit about it? "load" sounds more like it's getting the latest version than "get" does.

**Comment by andrei-ionescu** [comment]
@sezruby I added [HybridScanForIcebergTest.scala](https://github.com/microsoft/hyperspace/pull/358/files#diff-4905e3cb43d3852943ec05af7f34eb6e2d48dc0b707e700a9516cdda897f8729). Please have another look.

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanForIcebergTest.scala
Comment (sezruby) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanForIcebergTest.scala:149]: super nit:
```suggestion
        // Filter Index and Parquet format source file can be handled with 1 LogicalRelation.
```

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanForIcebergTest.scala
Comment (sezruby) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanForIcebergTest.scala:99]: Sorry I didn't revised this comment after refactoring 😅.

```suggestion
    // This test is for verifying plan transformation if appended files could be load with index
```

**Review by sezruby — COMMENTED** [review]
@andrei-ionescu Thanks! Seems it works as expected! 👍👍  
For start & end snapshot id test, I'd just like to check the use case & how those configs work well (or not well) with Hyperspace.

**Comment by andrei-ionescu** [comment]
@sezruby 

> For start & end snapshot id test, I'd just like to check the use case & how those configs work well (or not well) with Hyperspace.

Is there such check for Delta so that I could inspire from?

**Comment by sezruby** [comment]
@andrei-ionescu No, not sure Delta has the feature. 
How about this?

- build test data
  - snapshot 1) add a row with value 1 (append 1 file)
  - snapshot 2) add a row with value 2 (append 1 file)
  - snapshot 3) add a row with value 3 (append 1 file)
  - snapshot 4) add a row with value 4 (append 1 file)
  - snapshot 5) add a row with value 5 (append 1 file)

- build index with df with start-snapshot-id 2, end-snapshot-id 4 
  - query with df with start-snapshot-id 2, end-snapshot-id 4 w/o hybrid scan
  - query with df with start-snapshot-id 3, and check w/ hybrid scan
  - query with df with end-snapshot-id 3, and check w/ hybrid scan

**Comment by andrei-ionescu** [comment]
@sezruby I'll try to see it is possible but from my knowledge on Iceberg I don't think this kind of test is possible. The snapshot ids are used for time travel and to isolate date at read time. A snapshot is a sort of full image of the dataset -- is not a delta from the previous. This is a bit different than Delta.

Delta has the same parameter under `versionAsOf`. If there are tests to validate the functionality of Delta with time travel I'll do my best add it to Iceberg too although the implementation is a bit different.

Otherwise I would suggest to have a separate PR with time travel for Iceberg.

**Comment by sezruby** [comment]
@andrei-ionescu I guess `versionAsOf` of Delta Lake is the same as Iceberg only with `end-snapshot-id`.  
But both `start-snapshot-id` or `end-snapshot-id` handles "delta" dataset between two snapshots, right?

Hybrid scan will work w/o any link of version info - it utilizes the list of source files from DataFrame.

For time travel query optimization, to pick a proper version of index (after some refresh),
I added version history info to link delta version and index version in #272.

**Comment by andrei-ionescu** [comment]
@sezruby: I refactored it a bit and it is no longer needed to have knowledge of `start-snapshot-id`, `end-snapshot-id` and other internal Iceberg properties.

**Comment by sezruby** [comment]
LGTM thanks @andrei-ionescu!

**Comment by andrei-ionescu** [comment]
@rapoth, @imback82, @sezruby: The PR has 2 `LGTM`s already - thanks @imback82 & @sezruby. What's the process? What next steps are ahead to have it merged and complete the feature?

**Comment by imback82** [comment]
Let me take a final look since there were some changes since my last LGTM.

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:126]: Can you update this based on the implementation below?

**Review by imback82 — APPROVED** [review]

**Comment by imback82** [comment]
@andrei-ionescu Can you do a follow up PR to address https://github.com/microsoft/hyperspace/pull/358#discussion_r580425973? Thanks!

**Comment by andrei-ionescu** [comment]
@imback82: I create #362 to address the doc comment. 

@imback82, @sezruby, @rapoth: Thanks for help with this PR & feature.

**Comment by rapoth** [comment]
It has been a fantastic collaboration so far and I have personally learned a lot in the process! Thank you @andrei-ionescu! 🙂

**Comment by sezruby** [comment]
Thanks for the work @andrei-ionescu!

Could you also simply update the user document if possible? 
ref: https://microsoft.github.io/hyperspace/docs/ug-supported-data-formats/ 
https://github.com/microsoft/hyperspace/blob/master/docs/_docs/08-ug-supported-data-formats.md

**Pull request merged**
